Add pixelshuffle layers

kohr-h opened the pull request:
## Description ##
Pixel shuffle layers, as discussed in #13548 

## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

Closes #13548

kohr-h commented on python/mxnet/gluon/contrib/nn/basic_layers.py:
```diff
@@ -235,3 +235,135 @@ def _get_num_devices(self):
+        """Perform pixel-shuffling on the input."""
+        f1, f2, f3 = self._factors
+                                                              # (N, f1*f2*f3*C, D, H, W)
+        x = F.reshape(x, (0, -4, -1, f1 * f2 * f3, 0, 0, 0))  # (N, C, f1*f2*f3, D, H, W)
```
Does this always work? It generates a temporary 6D tensor, on CPU it's okay but I don't know about GPU.

ThomasDelteil replied:
Adding tests will let you test that automatically on all our supported platform when your PR is being tested in the CI. GPU / CPU / Windows / Linux / arm / etc.

ThomasDelteil commented:
@kohr-h, thanks for the contribution! Can you add tests for your new layers? Also can you add a bit more documentation and reference the original paper, thanks!

Roshrini commented:
@mxnet-label-bot Update [Operator, pr-awaiting-response]

kohr-h commented:
Axis swapping issue is fixed now. The 2D version now does the same as the PyTorch implementation. I'll add some more documentation.

kohr-h commented:
Just updated the superresolution example with the new layer, seems to work fine. The old download link wasn't working any more, so I switched to a GitHub mirror of the newer version of the dataset.

kohr-h commented:
Ready for review now.

Roshrini commented:
@apeforest @zhreshold @eric-haibin-lin Can you please help in reviewing this PR? Thanks
@mxnet-label-bot Update [pr-awaiting-review, Operator]

sandeep-krishnamurthy commented:
@ThomasDelteil - Can you please help review this PR?

vandanavk commented:
LGTM.

@zhreshold @Mut1nyJD for review.

@kohr-h if merging this PR would fix https://github.com/apache/incubator-mxnet/issues/13548, please add "Fixes https://github.com/apache/incubator-mxnet/issues/13548" in the PR description. this will automatically close the issue when the PR is merged.

vandanavk commented:
@mxnet-label-bot add [Gluon]

kohr-h commented:
@vandanavk I added the corresponding text.

ankkhedia commented:
@ThomasDelteil Could you please help review/merge this PR?

kohr-h commented:
:+1:
